packages/react/src/components/validation/components/validator.tsx

```diff
@@ -0,0 +1,76 @@
+import { ValidationResult, ValidationRule } from '@devextreme/interim';
+import {
+  PropsWithChildren, useContext, useEffect, useMemo, useRef,
+} from 'react';
+import { EditorContext } from '../../common/contexts/editor-context';
+import { ValidationEngineContext } from '../contexts/validation-engine-context';
+import { ValidationGroupContext } from '../contexts/validation-group-context';
+import { ValidatorContext } from '../contexts/validator-context';
+import { ValidationGroupId } from '../types';
+
+interface ValidatorImpl {
+  validate: () => ValidationResult,
+  validationRules?: ValidationRule[],
+  reset: () => void,
+  on: (event: string, callback: () => void) => void,
+  off: (event: string, callback: () => void) => void,
+}
+
+interface ValidatorProps extends PropsWithChildren {
+  validationGroup?: ValidationGroupId,
+  validateOnValueChange?: boolean
+}
+
+export function Validator({
+  validationGroup: validationGroupProp,
+  validateOnValueChange,
+  children,
+}: ValidatorProps) {
+  const validationGroupContext = useContext(ValidationGroupContext);
+  const editorContext = useContext(EditorContext);
+  const validationEngine = useContext(ValidationEngineContext);
+  const rules = useRef<ValidationRule[]>([]);
+  const performValidation = () => {
+    if (editorContext) {
+      const { editorName, editorValue, setEditorErrors } = editorContext;
+      const validationResult = validationEngine.validate(editorValue, rules.current, editorName);
+      setEditorErrors(
+        validationResult.isValid
+          ? [] : validationResult.brokenRules.map((rule: ValidationRule) => rule.message),
+      );
+      return validationResult;
+    }
+    return { isValid: true };
+  };
+
+  const validator = useRef<ValidatorImpl>({
+    validationRules: rules.current,
```

Comment:
Question.

Maybe I missed something, but when do the ```validationRules``` updates?

Reply:
Currently, rules register themselves only once on render, the update is unnecessary in this case.